babel/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -565,6 +565,111 @@ SqlCreate SqlCreateTable(Span s, boolean replace) :
     }
 }
 
+SqlCreate SqlCreateFunctionSqlForm(Span s, boolean replace) :
+{
+    SqlIdentifier functionName = null;
+    List<SqlIdentifier> fieldNames = new ArrayList<SqlIdentifier>();
+    List<SqlDataTypeSpec> fieldTypes = new ArrayList<SqlDataTypeSpec>();
+    DeterministicType isDeterministic = DeterministicType.UNSPECIFIED;
+    ReactToNullInputType canRunOnNullInput = ReactToNullInputType.UNSPECIFIED;
+    SqlIdentifier specificFunctionName = null;
+    final SqlDataTypeSpec returnsDataType;
+    boolean hasSqlSecurityDefiner = false;
+    SqlLiteral tempNumeric;
+    int typeInt;
+    final SqlNode returnExpression;
+}
+{
+    <FUNCTION>
+    functionName = CompoundIdentifier()
+    <LPAREN>
+    [
+      FieldNameTypeCommaListWithoutOptionalNull(fieldNames, fieldTypes)
+    ]
+    <RPAREN>
+    <RETURNS>
+    returnsDataType = DataType()
+    <LANGUAGE> <SQL>
+    (
+      <NOT> <DETERMINISTIC>
+      {
+        isDeterministic = DeterministicType.NOTDETERMINISTIC;
+      }
+    |
+      <DETERMINISTIC>
+      {
+        isDeterministic = DeterministicType.DETERMINISTIC;
+      }
+    |
+      <RETURNS> <NULL> <ON> <NULL> <INPUT>
+      {
+        canRunOnNullInput = ReactToNullInputType.RETURNSNULL;
+      }
+    |
+      <CALLED> <ON> <NULL> <INPUT>
+      {
+        canRunOnNullInput = ReactToNullInputType.CALLED;
+      }
+    |
+      <SPECIFIC>
+      {
+        specificFunctionName = CompoundIdentifier();
+      }
+    )*
+    [
+      <SQL> <SECURITY> <DEFINER>
+      {
+        hasSqlSecurityDefiner = true;
+      }
+    ]
+    <COLLATION> <INVOKER> <INLINE> <TYPE> tempNumeric = UnsignedNumericLiteral() {
+        typeInt = tempNumeric.getValueAs(Integer.class);
+        if (typeInt != 1) {
+            throw SqlUtil.newContextException(getPos(),
+                RESOURCE.numberLiteralOutOfRange(String.valueOf(typeInt)));
+        }
+    }
+    <RETURN> returnExpression = Expression(ExprContext.ACCEPT_SUB_QUERY)
+    {
+        return new SqlCreateFunctionSqlForm(s.end(this), replace,
+            functionName, specificFunctionName, fieldNames, fieldTypes,
+            returnsDataType, isDeterministic, canRunOnNullInput,
+            hasSqlSecurityDefiner, typeInt, returnExpression);
+    }
+
+
+}
+
+
+/**
+* Parse a "name1 type1 , name2 type2 ..." list,
+* the field type default is not nullable.
+*/
+void FieldNameTypeCommaListWithoutOptionalNull(
+        List<SqlIdentifier> fieldNames,
+        List<SqlDataTypeSpec> fieldTypes) :
```

Comment:
I believe SqlIdentifier is a specific type of SqlNode, and SqlIdentifier covered the usecases I needed here. Also, this function is more or less a copy of [FieldNameTypeCommaList](https://github.com/googleinterns/calcite/blob/05376d6bff11ccef690addb4f9c40ab16dd20bec/core/src/main/codegen/templates/Parser.jj#L5082) (excluded the optional null token). The field name could either be a simpleIdentifier or a compoundIdentifier. Could you give more reasoning as to why a SqlNode type would be more suitable here?

Reply:
I'm not saying change the type to SqlNode but to to use the SqlNodeList class instead of List<SqlIdentifer> and List<SqlDataTypeSpec>. As you said SqlIdentifier is already a SqlNode so it should work. Though if you are just copying an existing function it's probably best to keep it as is.

Reply:
I see, you mean the data type SqlNodeList would be useful here compares to the List of SqlIdentifier here. That is a good point. Since this is a modified copy of an existing function, I'll refrain from making more changes than needed unless there is more people prefer SqlNodeList than a list of SqlIdentifier here.

Reply:
Updated to use SqlNodeList in https://github.com/googleinterns/calcite/pull/79/commits/55b082e08f165ae803e9ebfc6c1bfa2a29e2b37d.